dbus/mako.c

```diff
@@ -115,6 +129,55 @@ static int handle_list_notifications(sd_bus_message *msg, void *data,
 			return ret;
 		}
 
+		ret = sd_bus_message_append(reply, "{sv}", "id",
+			"u", notif->id);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = sd_bus_message_open_container(reply, 'e', "sv");
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = sd_bus_message_append_basic(reply, 's', "actions");
+		if (ret < 0) {
+			return ret;
+		}
+
+		// Actions are an array of strings.
+		ret = sd_bus_message_open_container(reply, 'v', "as");
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = sd_bus_message_open_container(reply, 'a', "s");
+		if (ret < 0) {
+			return ret;
+		}
+
+		struct mako_action *action;
+		wl_list_for_each(action, &notif->actions, link) {
+			ret = sd_bus_message_append(reply, "s", action->key);
```

Comment:
Should we keep this an array of keys, or should we also add the title too?

Reply:
I changed it to a dict of identifier -> title. I think I finally fully understand dbus types now. Incidentally, the fact that our json output appears to have an extra level of array nesting (like we had a confusing discussion about in IRC a while back) seems to be a busctl display bug, gdbus does not show the same outer structure.

Reply:
Makes sense.

Clients can also add additional metadata to actions, like action-icons. Should we make actions a dict of identifier -> dict of action properties?

Reply:
w/e, let's just break our API if we need to.

Reply:
Ah sorry I was busy yesterday. We'll just have to break it before we release again. :)

Reply:
Oh, no worries! Just wanted to get this merged, no big deal to merge as-is :)